File: App/Java/app/src/main/java/com/example/tonjies/templateproject/module/select/presenter/SelectPresenter.java
package com.example.tonjies.templateproject.module.select.presenter;

import com.example.common.util.L;
import com.example.tonjies.templateproject.module.select.bean.City;
import com.example.tonjies.templateproject.module.select.bean.Country;
import com.example.tonjies.templateproject.module.select.bean.Province;
import com.example.tonjies.templateproject.module.select.model.SelectModel;
import com.example.tonjies.templateproject.module.select.presenter.contract.SelectContract;
import com.example.tonjies.templateproject.util.AStatic;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import okhttp3.ResponseBody;

/**
 * Created by 舍长 on 2019/1/5
 * describe:城市选择页面P层
 */
public class SelectPresenter implements SelectContract.SelectIPresenter {

    //View接口
    private SelectContract.SelectView selectView;

    //Model变量
    private SelectModel selectModel;

    //省份列表
    private List<Province> provinceList = new ArrayList<>();

    //城市列表
    private List<City> cityList = new ArrayList<>();

    //县列表
    private List<Country> countryList = new ArrayList<>();

    //用来返回，显示到界面上泛形为String类型的列表
    private List<String> dataList = new ArrayList<>();

    //当前选中的省份
    private Province selectProvince;

    //当前选中的城市
    private City selectCity;

    //当前选中的县
    private Country selectCountry;

    public SelectPresenter(SelectContract.SelectView selectView) {
        this.selectView = selectView;
        this.selectModel = new SelectModel();
    }


    /**
     * 查询对应省份的城市数据
     */
    @Override
    public void queryCities(int position) {
        if (position == -1) {

        } else {
            //获取当前选中省份
            selectProvince = provinceList.get(position);
        }
        //获取当前选中省份的标号
        int id = selectProvince.getId();
//        L.d("当前选中的省份是：" + selectProvince.getName() + " 省份的编号是:" + selectProvince.getId());
        queryFromServer(String.valueOf(id), "", AStatic.level_city);
    }

    /**
     * 查询当前城市的县级数据
     */
    @Override
    public void queryCountry(int position) {
        //获取当前选中的城市
        selectCity = cityList.get(position);
        //获取当前选中的城市编号
        int cityCode = selectCity.getId();
        //获取当前选中的省份编号
        int provinceCode = selectProvince.getId();
//        L.d("当前选中的城市是：" + selectCity.getName() + " 当前选中的城市编号是:" + cityCode + " 当前选中的省份编号是:" + provinceCode);
        queryFromServer(String.valueOf(provinceCode), String.valueOf(cityCode), AStatic.level_county);
    }

    //查询
    @Override
    public void queryCountryWeatherId(int position) {
        selectCountry=countryList.get(position);
        selectView.setWeatherId(selectCountry.getWeather_id());
    }

    /**
     * 查询服务器信息封装类，返回相应的数据
     */
    @Override
    public void queryFromServer(String provinceCode, String cityCode, final int type) {
        //请求地点数据
        selectModel.getCity(provinceCode, cityCode)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<ResponseBody>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(ResponseBody responseBody) {
                        try {
                            String string = responseBody.string();
                            switch (type) {
                                //当前请求的是全国省份的数据
                                case AStatic.level_province:
                                    setProvince(string);
                                    break;
                                //当前请求的是全国城市的数据
                                case AStatic.level_city:
//                                    L.d("" + string);
                                    setCities(string);
                                    break;
                                //当前请求的是全国县级别的数据
                                case AStatic.level_county:
//                                    L.d("当前请求的是全国县级别的数据" + string);
                                    setCountry(string);
                                    break;
                            }
//                            L.d("onNext" + string);
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        L.d("onError:" + e.toString());
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    /**
     * 请求的是省份
     */
    private void setProvince(String string) {
        try {
            //获得省份列表
            JSONArray jsonArray = new JSONArray(string);
            provinceList.clear();//清空之前省份列表的数据，防止数据重复加载
            dataList.clear();
            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject jsonObject = jsonArray.getJSONObject(i);
                //设置省份对象
                Province province = new Province(jsonObject.getInt("id"), jsonObject.getString("name"));

                provinceList.add(province);
                //添加到省份列表集合中
                dataList.add(province.getName());
            }
            selectView.setData(dataList, AStatic.level_province, "中国");
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    /**
     * 设置城市信息
     */
    public void setCities(String cities) {
//        L.d("获取到的城市信息:" + cities);
        try {
            JSONArray jsonArray = new JSONArray(cities);
            cityList.clear();
            dataList.clear();
            //遍历获取到的城市列表集合，将城市实体类添加到城市列表，将城市名添加到数据列表中
            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject jsonObject = jsonArray.getJSONObject(i);
                //设置城市对象
                City city = new City(jsonObject.getInt("id"), jsonObject.getString("name"));
                cityList.add(city);
                dataList.add(city.getName());
            }
            //String列表数据，级别为城市，省份名称
            selectView.setData(dataList, AStatic.level_city, selectProvince.getName());
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    /**
     * 设置县级
     */
    public void setCountry(String string) {
//        L.d("获取到的县级数据" + string);
        try {
            JSONArray jsonArray = new JSONArray(string);
            dataList.clear();
            countryList.clear();
            for (int i = 0; i < jsonArray.length(); i++) {
                //获取Object对象
                JSONObject jsonObject = jsonArray.getJSONObject(i);
                //设置县对象
                Country country = new Country(jsonObject.getInt("id"),
                        jsonObject.getString("name"), jsonObject.getString("weather_id"));
                countryList.add(country);
                dataList.add(country.getName());
            }
//            L.d("------------------------");
            selectView.setData(dataList, AStatic.level_county, selectCity.getName());
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    //销毁View对象
    public void detachView() {
        selectView = null;
    }
}

File: App/kotlins/app/src/main/java/com/example/tonjies/weatherx/module/select/presenter/SelectPresenter.kt
package com.example.tonjies.weatherx.module.select.presenter

import com.example.tonjies.weatherx.base.BasePresenter
import com.example.tonjies.weatherx.ext.BaseSubscriber
import com.example.tonjies.weatherx.ext.exexute
import com.example.tonjies.weatherx.module.select.bean.City
import com.example.tonjies.weatherx.module.select.bean.County
import com.example.tonjies.weatherx.module.select.bean.Province
import com.example.tonjies.weatherx.module.select.view.SelectView
import com.example.tonjies.weatherx.net.NetUtil
import com.example.tonjies.weatherx.util.Config
import com.example.tonjies.weatherx.util.L
import okhttp3.ResponseBody
import org.json.JSONArray
import java.util.ArrayList

class SelectPresenter : BasePresenter<SelectView>() {

    //省列表
    private val provinceList = ArrayList<Province>()

    //城市列表
    private val cityList = ArrayList<City>()

    //县列表
    private val countryList = ArrayList<County>()

    //返回数据列表
    private val dataList = ArrayList<String>()

    //请求级别
    private var LEVEL_PROVINCE = 0//省
    private var LEVEL_CITY = 1//市
    private var LEVEL_COUNT = 2//县

    //选中的省份
    private var selectProvince: Province? = null

    //选中的城市
    private var selectCity: City? = null

    //选中的县
    private var selectCountry: County? = null

    /**
     * 返回当前选中的县的天气编号
     */
    fun queryCoutry(position: Int) {
        //获取当前选中的县
        selectCountry = countryList[position]
        mView.setWeatherId(selectCountry!!.weatherId)
    }

    /**
     * 查询当前城市县
     */
    fun queryCountries(position: Int) {
        //获取当前选中城市
        selectCity = cityList[position];
//        mView.setCities(selectCity!!)

        //获取城市编号
        val cityCode = selectCity!!.cityCode
        //获取省份编号
        val provinceCode = selectProvince!!.provinceCode
        L.d("当前选中的城市是${selectCity!!.cityName},城市编号是$cityCode,省份编号是$provinceCode")
        queryFromServer(provinceCode.toString(), cityCode.toString(), "country")
    }

    /**
     * 查询当前省份城市列表
     */
    fun queryCities(position: Int) {
        //获取当前选中省份
        selectProvince = provinceList[position]
//        mView.setCities(selectProvince!!)
//        Config.provinceCode = position
        //获取省份编号
        val provinceCode = selectProvince!!.provinceCode
        L.d("当前选中的省份是${selectProvince!!.provinceName},省份的编号是${selectProvince!!.provinceCode}")
        queryFromServer(provinceCode.toString(), "", "city")
    }

    /**
     * 查询服务器信息封装类，返回相应的数据
     */
    fun queryFromServer(provinceCode: String, cityCode: String, type: String) {
        var netUtil = NetUtil()
        L.d("省份id是$provinceCode,城市id是$cityCode")
        netUtil.getCity(provinceCode, cityCode)
                .exexute(object : BaseSubscriber<ResponseBody>() {
                    override fun onNext(t: ResponseBody) {
                        val s: String = t.string()
                        L.d(s)
                        when (type) {
                            "province" -> {
                                L.d("当前选择的是省份")
                                setProvince(s)
                            }
                            "city" -> {
                                L.d("当前选择的是城市")
                                setCities(s)
                            }
                            "country" -> {
                                L.d("当前选择的是县")
                                setCoutries(s)
                            }
                        }
                    }

                    override fun onError(e: Throwable) {
                        L.d("${e.message}")
                    }
                })
    }

    /**
     * 设置县
     */
    private fun setCoutries(s: String) {
        L.d("获取的县级Json数据$s")
        val allCountries = JSONArray(s)
        countryList.clear()
        for (i in 0 until allCountries.length()) {
            val jsonObject = allCountries.getJSONObject(i)
            val country = County(jsonObject.getString("name"), jsonObject.getInt("id"), jsonObject.getString("weather_id"))
            countryList.add(country)
            dataList.clear()
            for (country in countryList) {
                dataList.add(country.name)
            }
            mView.setData(dataList, LEVEL_COUNT, selectCity!!.cityName)
        }
    }

    /**
     * 设置城市
     */
    private fun setCities(string: String?) {
        L.d("获取到的城市信息$string")
        val allCities = JSONArray(string)//解析获取到的城市Json数据
        cityList.clear()//清除城市列表
        //将获取到的城市添加到城市集合中，然后将每个城市的名字提取出来，赋值给dataList集合
        for (i in 0 until allCities.length()) {
            val jsonObject = allCities.getJSONObject(i)
            val city = City(jsonObject.getString("name"), jsonObject.getInt("id"))
            cityList.add(city)
            dataList.clear()
            for (city in cityList) {
                dataList.add(city.cityName)
            }
            //
            mView.setData(dataList, LEVEL_CITY, selectProvince!!.provinceName)
        }
    }

    /**
     * 设置省份
     */
    private fun setProvince(string: String?) {
        L.d("-------")
        val allProvinces = JSONArray(string)
        provinceList.clear()
        for (i in 0 until allProvinces.length()) {
            val province = allProvinces.getJSONObject(i)
            val provinces = Province(province.getString("name"), province.getInt("id"))
            provinceList.add(provinces)

            dataList.clear()

            for (province in provinceList) {
                dataList.add(province.provinceName)
            }

            mView.setData(dataList, LEVEL_PROVINCE, "中国")
        }
    }
}